pkg/controller/job/job_controller.go

```diff
@@ -1016,8 +1032,9 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 			// This pod was processed in a previous sync.
 			continue
 		}
+
 		considerPodFailed := isPodFailed(pod, jobCtx.job)
-		if podutil.IsPodTerminal(pod) || considerPodFailed || jobCtx.finishedCondition != nil || jobCtx.job.DeletionTimestamp != nil {
+		if canRemoveFinalizer(logger, jobCtx, pod, considerPodFailed) {
```

Comment:
if too many pods failed after the last sync, what prevents us from sending more than 500 pod patches in one sync?

That's the intent of `len(newSucceededIndexes)` below.

Reply:
This gives me an idea:
Make the job controller assume that `backoffLimit` could be nil, starting from this release, even if the API doesn't allow it.

We would skip calculating `status.failed` as a pod failure counter and only track `status.failedIndexes`, and instead  `failed=len(failedIndexes)`. Then, we can save an API call to update `uncountedTerminatedPods`, just like we do for succeded Pods in Indexed Mode.

Then, we can make the case in beta to allow nil `backoffLimit`.

Reply:
I think this idea is worth considering, but I'm wondering, that a semantic change of the `status.failed` might be problematic for some users.

Reply:
After a second thought about this I'm still not convinced.

First, using the `.status.failed` as a counter for failed indexes, may hide the information about individual pod failures, which might be important. For example, a user may tolerate failures with backoffs, but uses the `.status.Failed` field to mark jobs with at least one failure for investigation for the reason of flakyness. 

Second, I'm not sure the performance improvement is that essential. It only optimizes the unhappy path.

Another potential idea is to leave the meaning of `status.failed` field, but make it best effort when `backoffLimitPerIndex` is used. Then it could be under- or over- count in case of PodGC, but it would only matter for users using low `backoffLimit`.

Still, adding this as a point to reevaluate for Beta for now: https://github.com/kubernetes/enhancements/pull/4123/files. wdyt?

Reply:
> Second, I'm not sure the performance improvement is that essential. It only optimizes the unhappy path.

That's a good point. Also, for highly parallel jobs, the improvement is not substantial, as it's just one less API call for up to 500 pods. I anticipate that `backoffLimitPerIndex` is a feature that would mostly be used with high levels of parallelism.

Yes, let's leave this discussion to be re-evaluated for beta.